# Overview
Throughout this assignment, you will be performing certain well-defined tasks that’ll not only strengthen your concepts of Plotly and Dash, but will also help you learn a number of new concepts that are useful in analyzing, summarizing and visualizing data in the real world. 

Here is a template notebook with all the tasks mentioned in detail. **Please complete the tasks within the designated section only.**


## Task 1: Data Loading and Data Aggregation
* Load the 3 data files into the variables data_18, data_19, data_20. 

* Data aggregation is the process of gathering data and presenting it in a summarized format. The data may be gathered from multiple data sources with the intent of combining these data sources into a summary for data analysis.         
Similar to how this dataset involves 3 data files, you’ll often be working on combining information from 2 or more files and analysing it. More often than not, GroupBy is a very useful tool for this purpose. 

  Go through this article to learn more some helpful aggregation tools in Python: https://www.bmc.com/blogs/pandas-group-merge-concatenate-join/ 

  **You don't need to aggregate/ merge the datasets in this assignment, it is only for reading purposes.**

In [1]:
import pandas as pd 
data_18=pd.read_csv('https://raw.githubusercontent.com/dphi-official/Datasets/master/IT_Salary_Survey_EU_18-20/Survey_2018.csv')  
data_19=pd.read_csv('https://raw.githubusercontent.com/dphi-official/Datasets/master/IT_Salary_Survey_EU_18-20/Survey_2019.csv')  
data_20= pd.read_csv('https://raw.githubusercontent.com/dphi-official/Datasets/master/IT_Salary_Survey_EU_18-20/Survey_2020.csv') 

## Task 2: Data Analysis
* Display the first 5 rows of the 2018 survey data
* Display a concise summary of the 2020 data and list out 3 observations/inferences that you observe from the result. For this you will need to use the info() method.
* Display the descriptive statistics of the 2018 survey data
* Display the number of missing values in each column of the 2018 survey data
How many people responded to the survey in each of the 3 years? Has the number increased or decreased over the years?
* Display all the unique values and their frequency in the column - “Number of vacation days” of 2020 data. Write down your observations (at least one) for this result. 


In [2]:
# first five rows of 2018 survey
data_18.head()

Timestamp   Age  ... Company size   Company type
0  14/12/2018 12:41:33  43.0  ...     100-1000        Product
1  14/12/2018 12:42:09  33.0  ...       50-100        Product
2  14/12/2018 12:47:36  32.0  ...        1000+        Product
3  14/12/2018 12:50:15  25.0  ...        1000+        Product
4  14/12/2018 12:50:31  39.0  ...     100-1000  Ecom retailer

[5 rows x 14 columns]

In [3]:
# summary of 2020 survey 
data_20.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1253 entries, 0 to 1252
Data columns (total 23 columns):
 #   Column                                                                                                                   Non-Null Count  Dtype  
---  ------                                                                                                                   --------------  -----  
 0   Timestamp                                                                                                                1253 non-null   object 
 1   Age                                                                                                                      1226 non-null   float64
 2   Gender                                                                                                                   1243 non-null   object 
 3   City                                                                                                                     1253 non-null   o



*   The data types are either string(object) or float 
*   All columns except Timestamp,City ,Yearly brutto salary have missing data
*   Memory usage is 225.3 KB
*   





In [4]:
# statistical description
data_18.describe(include='all')

Timestamp         Age  ... Company size Company type
count                   765  672.000000  ...          750          730
unique                  761         NaN  ...            5           49
top     14/12/2018 12:53:47         NaN  ...     100-1000      Product
freq                      2         NaN  ...          260          451
mean                    NaN   32.183036  ...          NaN          NaN
std                     NaN    5.107268  ...          NaN          NaN
min                     NaN   21.000000  ...          NaN          NaN
25%                     NaN   29.000000  ...          NaN          NaN
50%                     NaN   32.000000  ...          NaN          NaN
75%                     NaN   35.000000  ...          NaN          NaN
max                     NaN   60.000000  ...          NaN          NaN

[11 rows x 14 columns]

In [5]:
# checking missing values for each column in the 2018 data
data_18.isnull().sum()

Timestamp                               0
Age                                    93
Gender                                 14
City                                   29
Position                               28
Years of experience                    33
Your level                             22
Current Salary                         15
Salary one year ago                   169
Salary two years ago                  302
Are you getting any Stock Options?     23
Main language at work                  15
Company size                           15
Company type                           35
dtype: int64

In [6]:
# investigating the numner of entries ( rows) in each year 
print(data_18.shape)
print(data_19.shape)
print(data_20.shape)

(765, 14)
(991, 23)
(1253, 23)


The number of respondents increased yearly from 765 in 2018 through 991 in 2019 to 1253 in 2020

In [7]:
# unique values and frequencies 
data_20['Number of vacation days'].value_counts()

30                                              488
28                                              233
27                                              102
25                                               91
26                                               71
24                                               67
29                                               24
20                                               13
21                                               10
32                                                8
22                                                8
31                                                8
36                                                5
35                                                5
40                                                4
23                                                4
0                                                 4
33                                                3
45                                                3
14          

233 people took 28 days vacation while 1 person took 34 days vacation

## Task 3: Data Cleaning
* Rename the column ‘Position ‘ in the 2020 data to ‘Position’. (without the blank space)
* Check for missing values in 2020 data for all the columns. If there are no missing values, proceed to the next step. If there are missing values in the dataset,
  * For categorical variables, fill the missing values with the mode of the data. Remember if the data type of any variable is ‘object’, it is categorical variable. 
  * For numerical variables, fill the missing values with the mean of the data.

Here's a good blog that displays multiple methods of filling (imputing) missing values: https://jamesrledoux.com/code/imputation 
* Drop the timestamp column for all the three years data since the date and time at which a person filled the survey is irrelevant to us. The year matters and we already know that from the dataset’s name.
* Perform any other data cleaning steps you believe are necessary. (removing outliers, handling missing values in a way to beautify visualizations, making the categories uniform i.e python and Python should mean the same thing etc.) Note that the same steps will have to be performed for all 3 data files.

In [8]:
#renaming the column 'Position '
data_20=data_20.rename(columns={'Position ':'Position'})

In [9]:
#checking for missing values in 2020 data
data_20.isnull().sum()

Timestamp                                                                                                                    0
Age                                                                                                                         27
Gender                                                                                                                      10
City                                                                                                                         0
Position                                                                                                                     6
Total years of experience                                                                                                   16
Years of experience in Germany                                                                                              32
Seniority level                                                                                                

In [10]:
#filling missing values for numerical and categorical columns(cell below)
import numpy as np

num_cols = data_20.select_dtypes(include=np.number).columns      # getting all the numerical columns

data_20[num_cols] = data_20[num_cols].fillna(data_20[num_cols].mean())

In [11]:
cat_cols = data_20.select_dtypes(include = 'object').columns    # getting all the categorical columns

data_20[cat_cols] = data_20[cat_cols].fillna(data_20[cat_cols].mode().iloc[0])

In [12]:
# upon investigation data_19 does not have the time stamp column .So we shall only drop from the other two survey datas
data_18=data_18.drop(['Timestamp'],axis=1)


In [13]:
data_20=data_20.drop(['Timestamp'],axis=1)

In [14]:
# checking effectively that we have dropped the timestamp column
data_18.info()
data_19.info()
data_20.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 765 entries, 0 to 764
Data columns (total 13 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Age                                 672 non-null    float64
 1   Gender                              751 non-null    object 
 2   City                                736 non-null    object 
 3   Position                            737 non-null    object 
 4   Years of experience                 732 non-null    float64
 5   Your level                          743 non-null    object 
 6   Current Salary                      750 non-null    float64
 7   Salary one year ago                 596 non-null    float64
 8   Salary two years ago                463 non-null    float64
 9   Are you getting any Stock Options?  742 non-null    object 
 10  Main language at work               750 non-null    object 
 11  Company size                        750 non-n

In [15]:
#performing data cleaning operations for 2019 and 2018
num_cols = data_18.select_dtypes(include=np.number).columns      # getting all the numerical columns

data_18[num_cols] = data_18[num_cols].fillna(data_18[num_cols].mean())
num_cols = data_19.select_dtypes(include=np.number).columns      # getting all the numerical columns

data_19[num_cols] = data_19[num_cols].fillna(data_19[num_cols].mean())

In [17]:
cat_cols = data_18.select_dtypes(include = 'object').columns    # getting all the categorical columns

data_18[cat_cols] = data_18[cat_cols].fillna(data_18[cat_cols].mode().iloc[0])
cat_cols = data_19.select_dtypes(include = 'object').columns    # getting all the categorical columns

data_19[cat_cols] = data_19[cat_cols].fillna(data_19[cat_cols].mode().iloc[0])

In [18]:
# This is done in anticipation of the visualization section
data_19['Company type'].value_counts()

Product                 650
Startup                 181
Consulting / Agency     117
Bodyshop / Outsource     30
University                6
Bank                      6
Outsource                 1
Name: Company type, dtype: int64

## Task 4: Data Visualization using Plotly
**Note:** All the tasks below need to be completed using only Plotly and no other Data Visualization library.

* Create a pie chart to analyze the Company types in the year 2019. Are Consulting / Agency companies more popular than Startups? 
* Create a line plot of the Total years of experience vs the current salary(taking the median salary for each of the different experience years) of the year 2018.
* Now, create the above plot again and add 2 more line plots to the same graph, that display the Total years of experience vs the median Yearly brutto salary (without bonus and stocks) of the year 2019 and 2020.
* Create a bar chart to analyse the popularity of the main technology/ programming languages amongst the respondents in the year 2020. Which technology is the most popular? Which technology is the least popular (with less than 4 responses)?
* Create a pie plot indicating the gender ratio of the respondents in the year 2020.


In [19]:
import plotly.express as px
df = data_19['Company type']
di=[650,181,117,30,6,6,1]
fig = px.pie(df,values=di, names=['Product','Startup','Consulting / Agency','Bodyshop / Outsource','University','Bank','Outsource'], title='Company types in the year 2019')
fig.show()

Consulting/agency are less popular than start ups 

In [20]:
data_18.head()

Age Gender     City  ... Main language at work  Company size   Company type
0  43.0      M  München  ...               Deutsch      100-1000        Product
1  33.0      F  München  ...               Deutsch        50-100        Product
2  32.0      M  München  ...               Deutsch         1000+        Product
3  25.0      M  München  ...               English         1000+        Product
4  39.0      M  München  ...               English      100-1000  Ecom retailer

[5 rows x 13 columns]

In [21]:
#getting the median salary for each year of experience
data_18.groupby('Years of experience').median()['Current Salary']

Years of experience
0.000000      50000.000000
0.500000      46000.000000
1.000000      46000.000000
1.500000      58000.000000
2.000000      50000.000000
2.500000      50000.000000
3.000000      55555.000000
4.000000      60000.000000
4.500000      39150.000000
5.000000      60000.000000
6.000000      64000.000000
7.000000      68000.000000
7.500000      65000.000000
8.000000      70000.000000
8.548497      68381.765333
9.000000      70000.000000
10.000000     70000.000000
11.000000     71000.000000
12.000000     71000.000000
13.000000     73500.000000
14.000000     71500.000000
15.000000     68000.000000
16.000000     90500.000000
17.000000     80000.000000
18.000000     72500.000000
19.000000     84000.000000
20.000000     80000.000000
21.000000     90000.000000
25.000000    109000.000000
30.000000     64000.000000
38.000000     58000.000000
Name: Current Salary, dtype: float64

In [22]:
ji=data_18['Years of experience'].unique()
p=np.sort(ji)

In [23]:
p

array([ 0.        ,  0.5       ,  1.        ,  1.5       ,  2.        ,
        2.5       ,  3.        ,  4.        ,  4.5       ,  5.        ,
        6.        ,  7.        ,  7.5       ,  8.        ,  8.54849727,
        9.        , 10.        , 11.        , 12.        , 13.        ,
       14.        , 15.        , 16.        , 17.        , 18.        ,
       19.        , 20.        , 21.        , 25.        , 30.        ,
       38.        ])

In [27]:
#line plot with plotly express
medians=[50000.0,46000.0,46000.0,58000.0,50000.0,50000.0,55555.0,60000.0,39150.0,60000.0,64000.0,68000.0,65000.0,70000.0,68381.765333,70000.0,70000.0,71000.0,71000.0,73500.0,71500.0,68000.0,90500.0,80000.0,72500.0,84000.0,80000.0, 90000.0,109000.0,64000.0,58000.0]
import plotly.express as px
df = data_18
fig = px.line(df, x=p, y=medians,labels={'x':'Total years of experience','y':'the current salary'}, title='Total years of experience vs the current salary')
fig.show()

In [32]:
data_19.groupby('Years of experience').median()['Yearly brutto salary (without bonus and stocks)'] 


Years of experience
0      55000.0
1      46800.0
2      52500.0
3      55000.0
4      62000.0
5      65000.0
6      68000.0
7      67000.0
8      74500.0
9      74500.0
10     76500.0
11     76500.0
12     77700.0
13     70000.0
14     80000.0
15     75000.0
16     72000.0
17     75000.0
18     82500.0
19     73000.0
20     78000.0
22    100500.0
24    115000.0
25     79800.0
28    110000.0
30     60000.0
Name: Yearly brutto salary (without bonus and stocks), dtype: float64

In [40]:
k=data_19['Years of experience'].unique()
q=np.sort(k)
medians19=[ 55000.0,46800.0,52500.0,55000.0,62000.0,65000.0,68000.0,67000.0,74500.0,74500.0,76500.0,76500.0,77700.0,70000.0,80000.0,75000.0,72000.0,75000.0,82500.0,73000.0,78000.0,100500.0,115000.0,79800.0,110000.0,60000.0]

In [63]:
# The years of experience above are not completely numerical ,this will be edited manually ,where we have lesss than year, I will replace with 0
# There are many more adjustments to be made ,so the editing will be done even to the data_20 frame and the medians will be recalculated
data_20['Total years of experience'].unique()


array([  5. ,   7. ,  12. ,   4. ,  17. ,   6. ,   8. ,  15. ,   2. ,
        25. ,  10. ,  14. ,  11. ,  18. ,  13. ,  30. ,   3. ,  40. ,
        26. ,  23. ,   9. ,  19. ,  20. ,   5.5,  22. ,  16. ,   0.8,
         1. ,   1.5,   6.5,  21. ,   7.5,   2.5,  28. ,  29. ,  24. ,
         0. ,   4.5,  27. ,  31. , 383. ,   3.5])

In [48]:
data_20['Total years of experience'].replace({'1 (as QA Engineer) / 11 in total': 1, '15, thereof 8 as CTO': 15,'6 (not as a data scientist, but as a lab scientist)':6,'less than year':0,'1,5':1.5,'2,5':2.5,}, inplace=True)
data_20['Total years of experience']

0        5
1        7
2       12
3        4
4       17
        ..
1248     9
1249    10
1250    15
1251     7
1252     2
Name: Total years of experience, Length: 1253, dtype: object

In [54]:
# converting to float since only numerical values can be plotted on the line graph
data_20['Total years of experience']=data_20['Total years of experience'].astype(float)

In [55]:
# computing the median for each total year of experience 
data_20.groupby('Total years of experience').median()['Yearly brutto salary (without bonus and stocks) in EUR'] 

Total years of experience
0.0       40000.0
0.8       29750.0
1.0       48500.0
1.5       49925.0
2.0       50000.0
2.5       60000.0
3.0       56000.0
3.5       62500.0
4.0       60000.0
4.5       67500.0
5.0       65000.0
5.5       48000.0
6.0       67750.0
6.5       58000.0
7.0       70000.0
7.5       68000.0
8.0       68000.0
9.0       74000.0
10.0      75000.0
11.0      75000.0
12.0      75000.0
13.0      83000.0
14.0      80000.0
15.0      78000.0
16.0      83000.0
17.0      76000.0
18.0      85500.0
19.0      87000.0
20.0      80000.0
21.0      80000.0
22.0      70000.0
23.0     100000.0
24.0     130000.0
25.0      75000.0
26.0      93000.0
27.0      73500.0
28.0      78500.0
29.0      28800.0
30.0      64500.0
31.0     110000.0
40.0      70000.0
383.0     70000.0
Name: Yearly brutto salary (without bonus and stocks) in EUR, dtype: float64

In [60]:
gi=data_20['Total years of experience'].unique()
ss=np.sort(gi)
ss

array([  0. ,   0.8,   1. ,   1.5,   2. ,   2.5,   3. ,   3.5,   4. ,
         4.5,   5. ,   5.5,   6. ,   6.5,   7. ,   7.5,   8. ,   9. ,
        10. ,  11. ,  12. ,  13. ,  14. ,  15. ,  16. ,  17. ,  18. ,
        19. ,  20. ,  21. ,  22. ,  23. ,  24. ,  25. ,  26. ,  27. ,
        28. ,  29. ,  30. ,  31. ,  40. , 383. ])

In [67]:
# the number of years 383.0 will be replaced by 38.3
lik=np.where(ss==383.0,38.3,ss)
fin=np.sort(lik)
fin

array([ 0. ,  0.8,  1. ,  1.5,  2. ,  2.5,  3. ,  3.5,  4. ,  4.5,  5. ,
        5.5,  6. ,  6.5,  7. ,  7.5,  8. ,  9. , 10. , 11. , 12. , 13. ,
       14. , 15. , 16. , 17. , 18. , 19. , 20. , 21. , 22. , 23. , 24. ,
       25. , 26. , 27. , 28. , 29. , 30. , 31. , 38.3, 40. ])

In [61]:
medians20=[40000.0,29750.0,48500.0,49925.0,50000.0,60000.0,56000.0,62500.0,60000.0,67500.0,65000.0,48000.0,67750.0,58000.0,70000.0,68000.0,68000.0,74000.0,75000.0,75000.0,75000.0,83000.0,80000.0,78000.0,83000.0,76000.0,85500.0,87000.0,80000.0,80000.0,70000.0,100000.0,130000.0,75000.0,93000.0,73500.0,78500.0,28800.0,64500.0,110000.0,70000.0,70000.0]

In [69]:
#multiple lin eplots with plotly graph objects
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# Create figure with secondary y-axis
fig = make_subplots(specs=[[{"secondary_y": True}]])

# Add traces
fig.add_trace(
    go.Scatter(x=p, y=medians, name="2018 data"),
    secondary_y=False,
)

fig.add_trace(
    go.Scatter(x=q, y=medians19, name="2019 data"),
    secondary_y=True,
)
fig.add_trace(
    go.Scatter(x=fin, y=medians20, name="2020 data"),
    secondary_y=True,
)

# Add figure title
fig.update_layout(
    title_text="Total years of experience versus current salary"
)

# Set x-axis title
fig.update_xaxes(title_text="Total years of experience")

# Set y-axes titles
fig.update_yaxes(title_text="<b>primary</b> current salary", secondary_y=False)
fig.update_yaxes(title_text="<b>secondary</b> current salary", secondary_y=True)
fig.update_yaxes(title_text="<b>secondary</b> current salary", secondary_y=True)

fig.show()

In [70]:
data_20['Your main technology / programming language'].value_counts()

Java                          311
Python                        164
PHP                            56
C++                            38
JavaScript                     34
                             ... 
Web                             1
--                              1
.net, c++, spss, embeddded      1
TypeScript, Kotlin              1
Sketch, Figma                   1
Name: Your main technology / programming language, Length: 256, dtype: int64

In [75]:
#creating a data frame from the valuecounts
# the next two cells below are for my personal learning since I had some doubts .I just realised that in the previous line plots I could create dataframes to extract
#the medians instead of manually writing them into a list  
df=pd.DataFrame(data_20['Your main technology / programming language'].value_counts())
df.head()
df.shape

(256, 1)

In [77]:
df['Your main technology / programming language']
df.columns

Index(['Your main technology / programming language'], dtype='object')

In [78]:
df.index

Index(['Java', 'Python', 'PHP', 'C++', 'JavaScript', 'Javascript', 'C#',
       'python', 'Scala', 'Swift',
       ...
       'nothing', 'JavaScript/TypeScript', 'Network', 'Nodejs', 'julia', 'Web',
       '--', '.net, c++, spss, embeddded', 'TypeScript, Kotlin',
       'Sketch, Figma'],
      dtype='object', length=256)

In [82]:
#plotting bar chart of main technologies
import plotly.express as px
fig = px.bar(df, x=df.index, y=df['Your main technology / programming language'])
fig.show()

In [83]:
#checking technologies with less than 4 responses
df[df['Your main technology / programming language']<4]

Your main technology / programming language
Cloud                                                                 3
Node.js                                                               3
Embedded                                                              3
React                                                                 3
c++                                                                   3
...                                                                 ...
Web                                                                   1
--                                                                    1
.net, c++, spss, embeddded                                            1
TypeScript, Kotlin                                                    1
Sketch, Figma                                                         1

[221 rows x 1 columns]

Java is the most popular technology with 311 responses while we have 221 texhnologies with less than 4 responses as shown in the above cell .Examples of such technologies are c++,React and Cloud

In [85]:
data_20['Gender'].value_counts()

Male       1059
Female      192
Diverse       2
Name: Gender, dtype: int64

In [88]:
#pie chart of Gender ratio in 2020
import plotly.express as px
dp = data_20['Gender']
fig = px.pie(dp,names=['Male','Female','Diverse'],values=[1059,192,2], title='Gender ratio in 2020')
fig.show()

## Bonus Section [Optional but carries bonus marks]
This dataset is as raw and real as it can get while conducting yearly surveys. You might have observed that the data is not clean and structured and requires some thorough cleaning before deriving meaningful plots. When combined with the power of Plotly and Dash, there are endless possibilities for the insightful visualizations you can create. 

This section is to let you experiment, explore and create as many visualizations as you’d like. You never know, if we like the creativity and the extra work, you might receive some bonus marks!


# Conclusion
This brings us to the end of the assignment and to the bootcamp. We hope you had a great learning time. :)

Now, you can submit your notebook for assessment. 